Allow representations to contain differentials

adrn opened the pull request:
This is a stepping stone PR towards adding support for proper motions and velocities in `astropy.coordinates`. It turns out, we need to either (1) allow `Representation`s to carry around `Differential`s, or (2) create a new container class for representation+differentials. Option 2 seems better from a design perspective, but would lead to some API-breaking changes (or else a lot of really annoying code to provide backwards-compatibility). As this only really affects the lower-level API (and not the `Frame` or `SkyCoord`-level API), we decided that, in the interest of time, we would make this "hack" for now and just clearly document that the low-level API could change.

This PR adds the necessary code to allow `Representation`'s to accept any number of `Differential`'s in their initializers. For simplicity, we are very strict:
* The `Differential` instances must have the same shape as the representation. (I could see loosening this a little and just requiring that they are broadcastable on initialization, but leave that as an open question)
* Shape-changing (e.g., `_apply()`-related operations) affect both the representation and the differentials

Check out this document if you're interested in seeing a very messy document that lays out some more details: https://docs.google.com/document/d/1dZIG7CG8h-C4lcegtx71eUc76K5LHtgRBsXk4iRa2ds/edit?usp=sharing

eteq commented:
(As a warning to others, @adrn and I have been doing a lot of out-of-band discussion about this so it might be difficult to follow the doc...)

On the questions:

> How do we modify the __repr__() on BaseRepresentation when there are differentials associated?

I'd vote that if there are no differentials, it should look just like the current version. E.g. (for comparison with the others below,
```
<CartesianRepresentation (x, y, z) in kpc
    ( 0.0663711,  0.75566441, -1.29113804)>
``` 

If differentials are present, the most straightforward option would be:
```
<CartesianRepresentation (x, y, z) in kpc
    ( 0.0663711,  0.75566441, -1.29113804),
 differentials=(<CartesianDifferential (d_x, d_y, d_z) in kpc
    ( 1.82124186,  1.48662901, -1.21030373)>,)>
```
which is pretty trivial to implement.

I could also see an argument for this, though:
```
<CartesianRepresentation (x, y, z) in kpc
    ( 0.0663711,  0.75566441, -1.29113804),
 differentils=(CartesianDifferential(...), )>
```
e.g., don't show the actual data, but expect the user to do ``.differentials`` if they want to see the data.  That prevents the repr from getting awkwardly long if there's lots of differentials.  I think I slightly favor this over the above, but either seem ok.

> Where should I document the subtleties of this? We want to document it somewhere, but also we don't really want people to use it explicitly, or at least have a clear warning on top. Should I just add a section to docs/coordinates/representations.rst near the bottom?

Yes, I like just what you suggested here.  But be sure to add a ``.. warning:: this code is experimental and may change in future versions!`` at the top of the section.

> One slightly annoying side effect of this is that any subclass of BaseRepresentation has to remember to pass through the differentials in to_cartesian() and from_cartesian() (see examples in this PR, e.g., SphericalRepresentation. Should we just accept that and document it?

Hmm, there is an alternative: have `to_cartesian`/`from_cartesian` be on `BaseRepresentation` and have the subclasses instead implement `_real_to_cartesian`/`_real_from_cartesian`, which does *not* ever include the differentials in its result.  Then `BaseRepresentation.to_cartesian` would be something like:
```
def to_cartesian(self):
    repr = self._real_to_cartesian()
    repr._differentials = self._differentials 
    return repr
```

adrn commented:
First two points: sounds good!

On the `to_cartesian`/`from_cartesian` idea: I was thinking something similar. My idea was to have the subclasses implement (something like) `_to_cartesian_components()`, which returns an `OrderedDict` with just the `x`, `y`, `z` data, and `_from_cartesian_components()`, which returns an `OrderedDict` with keys set to the component names of the given representation. I guess I was thinking about moving the actual creation of the object up to the superclass (since right now, they all do `copy=False` in the initializer as well.)

eteq commented:
@adrn: the commits I just pushed up add the `with_differential` method.  The last of the three actually does something different: it explicitly addresses passing in a string to `represent_as`.  That wasn't being accounted for, but on reflection I think it's best to do this explicitly so as to yield a more helpful error message.  But if you disagree feel free to back that out or do something different.

One other thing: I noticed the machinery for adding differentials is on `BaseRepresentationOrDifferential`, not `BaseRepresentation`.  I think that's not what we want: differentials attached to other differentials creates an ambiguity: if I have an acceleration, should it be ``rep.differential[0].differential[0]``, or ``rep.differential[0]``?  Also, it's impossible to do the re-representations anyway if you don't have the representation and all the lower-order differentials.  So I think differentials should only be attachable to representations, *not* other differentials.  (Note that if you change this, you'll also need to move the `BaseRepresentationOrDifferential._validate_differentials` method which I added in those last few commits.)

adrn commented:
@eteq ah, yea, I agree - I just moved the relevant code to `BaseRepresentation` locally. Will push some more changes shortly.

mhvk commented:
@adrn, @eteq - only high-level comments and questions in the first instance. From highest to lowers:
1. Question: Why is this best done inside `representations` rather than at the frame level? I ask mostly because it makes sense for coordinates to have just RV or PM, but not really as much for representations, which suggests that this doesn't belong in representations.
2. Comment: I do think we do need to get this right in one go: to have a "hack" just ahead of a LTS version means that we're stuck with the hack. Better it seems to bite the bullet and go for a design where we create a new class that holds both a representation and a differential. (I doubt that there are that many implementations of custom representations that we would break at this point.) [Badly thought out idea: could the new ones have the old names, and the old representations get new names? Does this even help?]
3. Comment: from some of the tests, it seemed mix-and-match representations and differentials were allowed (spherical and cylindrical); is this the case and if so was this indeed intended? I would suggest that for every case, it should be possible to use the differential with as a base the representation it is attached to (i.e., without further converting the base). So, a spherical representation could hold spherical, unispherical and radial differential (plus coslat variants), but a cylindrical representation only a cylindrical differential.
4. Comment: I feel operations on representations should not drop the differential but rather just error. (This allows them to start working *with* differentials later, for cases where that makes sense.)

adrn commented:
@mhvk @eteq 

1. My recollection is that when I looked into this idea, I decided that it would end up requiring a lot of detailed changes to the frame classes and duplicated code because of some unfortunate choices in the development of the frames. Right now, the representations are stored as `._data` on the frame classes (ugh), and most methods work with this attribute alone. We'd have to modify everything to also look for `._differentials` (or whatever). @eteq and I briefly discussed the possibility of creating a new container class for representations and differentials that would replace `._data`, but we're really down to the wire here and trying to carve out a minimal path towards getting something in before the feature freeze. 

2. I just don't think we have the time (or at least I don't) before next weekend. I do agree in principle and think that it would be the cleaner way to go, but I worry about the required fixes and downstream changes we'd need to make elsewhere for any API-breaking changes we make to the representations. Our compromise was to add the minimal functionality we needed and document it as experimental; wouldn't that mean we don't need to provide LTS for this functionality? In any case, I think we need to make a clear decision ASAP about this because we can't proceed with any of the further pull requests without this... 

3. Right now, yes, mix-and-match representations and differentials are allowed.

4. That's a fair point!

eteq commented:
@mhvk:
On 1. Fair point - I think I originally did test velocity implementations just as you propose, but eventually we realized attaching them to the representations is important for the reasons @adrn gave above, *and* because it makes certain aspects of velocity support easier to implement (@adrn and I are working on a couple of PRs for that now). More philosophically, after thinking and talking with a few folks about it, I think this actually *does* make more sense: basically, you can't do really anything with differentials (other than cartesian) *without* the representation (and any lower-order derivatives if we implement acceleration et al. in the future), so it makes as much sense to attach them to the representation as shared in the frames.

On 2. I disagree that we're doing a "quick hack" for LTS.  We've been thinking about this for over a year now and its time to get something in to support velocities.  The feature freeze is sort of acting as a stick so that the perfect isn't the enemy of the good. I also think it's not unreasonable to have this release include a warning in the docs that says "this might change in future versions" which is fairly standard for big changes like this.

As for whether there should be a container class with both representations and differentials, that was @adrn's initial idea, but I think I convinced him otherwise on the basis that this would be a major a API-breaking change, and @adrn realized it would be a *lot* of work with a variety of subtle consequences to change `._data` to be a container class (as @adrn said above about 1). And there *are* people I know of using this representation machinery (most notably Sunpy), which means there are probably more who we don't know about (and won't know about until the start using it and it's all borken).  So I think that ship has already sailed.

On 3. Good point... We ended up with the mix-and-match option because as long as its attached to the representation, the differential can always be re-cast into whatever representation is desired, so one can always get whatever one wants.  But sometimes you don't *want* to.  E.g. if you have a cartesian representation and differential and just want to get the ``.lat``, you shouldn't be forced to transform the differential as it's a performance waste.  On some level this is an argument *for* your point 1, but given the restrictions we described above, it seems like this is the lesser of the various evils.

(I'm ambivalent on 4)

mhvk commented:
While reviewing the affine transform, it would seem handy if for representations that have differentials, at least `to_cartesian` would return a `CartesianRepresentation` that also had the differentials attached, already transformed to `CartesianDifferential`.

adrn commented:
The main lingering issue with this PR is that the change in what `Representation` classes need to implement---from `to_cartesian()`/`from_cartesian()` to `_to_cartesian_helper()`/`_from_cartesian_helper()`---is API breaking as it stands. I searched GitHub for any subclasses of `BaseRepresentation` outside of Astropy forks, and I only found one instance: https://github.com/cta-observatory/ctapipe/blob/master/ctapipe/coordinates/representation.py

Seems that the two options then are:

1. Break the API, issue a pull request to the above repository to update the code.
2. Rename the current `BaseRepresentation` -> e.g., `NewBaseRepresentation`, and subclass `BaseRepresentation` from that to add the old functionality back. We could raise a deprecation warning if they use that class. From my view, the disadvantage of this is that then we're stuck with the new name of the class unless we deprecate `NewBaseRepresentation` at some point in the future...But that seems like adding more work. 

Thoughts? @mhvk @eteq

eteq commented:
One other pending item: I think that operations on representations with differentials shouldn't necessarily error.  I think the idiom of using representations as vectors is useful enough in the transformations that I'm concerned it's a notable API "break" to have representations with differentials suddenly start erroring.  For example, if I do this:
```
i = ICRS(0*u.deg, 0*u.deg, 1000*u.au,  
         pm_ra=0*u.marcsec/u.yr, pm_dec=0*u.marcsec/u.yr, 
         radial_velocity=0*u.km/u.s)
i.transform_to(GCRS)
````
I get an error in the current implementation.  We can fix that, of course, by dropping the differential or something, but the worry is that this has been used elsewhere and suddenly errors will start appearing... At the very least it seems like matching units should be able to be cartesian operated-on?

mhvk commented:
After sleeping on this for a night:
1. I think we should *not* use a list for the differentials -- the additional code is already assuming a velocity will be the first item, and surely any user code is going to do similarly, or worse, make up their own convention of order. That seems a recipe for problems if we want to extend it. I would suggest instead to use a `dict`, with a convention that an integer key is always a time derivative (with 0 for an offset, 1 for velocity, 2 for acceleration, etc.). I would then suggest to change the name to `differential` (i.e., not plural), so that, e.g., a velocity would be `rep.differential[1]`.
2. I'm still uncomfortable with allowing any differential, and would prefer a check that a given representation works as a base directly
3. I hope that it will be possible for at least `to_cartesian` to return a `CartesianRepresentation` with differentials attached. Then, ideally, `from_cartesian` gains a `differential_cls` argument for the reverse, and with that it is not much further to change also `represent_as` and `from_representation`, so that one could, e.g., use something like `r1.represent_as(SphericalRepresentation, differential_cls=SphericalCosLatDifferential)`.

mhvk commented:
@eteq - I don't think we should be forgiving for  "breaking" code for objects for which velocities have been added -- in that case it *should* break, as it cannot know the answer.

I think you are right, though, that some operations, such as adding/subtracting two Cartesian representations with associated differentials, are well defined. For multiplication and division, just applying the same multiplier to all differentials may also make sense (it is as least well-defined). Less sure about `.dot` and `.cross`, though. In any case, I think we should leave this for another PR; I already don't see that I can review all this today in the few hours I will have. (But hopefully an extension over the weekend is OK.)

mhvk commented:
@adrn - in the time I do have I hope to look into your API breaking issue, especially to see if perhaps it can be avoided still. If not, I'd vote for revisiting the idea of a class that holds both (which might be a subclass of `BaseRepresentation` -- needs thought).

adrn commented:
@mhvk On your 3 enumerated comments:

1. I'm a little confused with the idea that the 0th element should be an offset (as in, I just can't think of a use case for that) but I'd be willing to spec it out. We chose a list because we decided it probably wouldn't make sense to have a 2nd derivative without also the 1st derivative, 3rd without 1st and 2nd, etc. But working with a dictionary with integer keys might internally be easier than inserting into a list. I also don't think we should restrict to time derivatives, but then I don't know of any way to validate the input units - do you have thoughts? 

2. I would be ok with this. Is there an easy way for a differential to check if a given representation is a valid base?

3. `to_cartesian` does return a `CartesianRepresentation` with the differentials attached. `represent_as()` has already been updated to allow you to specify a list of classes (for the representation and all differentials), but I do like your idea of using a single kwarg instead.

mhvk commented:
@adrn - I hope to go through this with a finer comb tonight (as I write, 8 PM here in Stockholm), after having a high-level look at PM/radial-velocity. Let me know if this is not a good moment, as tomorrow is another option.

@eteq, @adrn - On the 3 higher-level points:

1. dict instead of list: `0` would be useful for pure offsets - a lot of my tests were written for that - I think it will be useful (but in any case, it's there mostly so the number has immediate sense). I think the numbers themselves should be time derivatives, but this doesn't preclude using the dict for other derivatives with a different name.

2. differentials checking base - yes and no, there is `_check_base` but this is too narrow and actually less well defined than I thought it was (e.g., for `RadialDifferential.to_cartesian`, a `RadialRepresentation` base is no use). Since time is running out for 2.0, perhaps the goal for now would be to write the code such that the differential always makes sense relative to the base.  See #6233 for an issue to remind us.

3. Great that this has already been done; hope to look at it later.

adrn commented:
Thanks @mhvk - you can start looking through #6219. From now, I'll be offline until ~midnight your time today, so I can respond to any comments you leave then. But, I don't expect you to be there to chat -- if possible, let's try to touch base tomorrow at some point?

StuartLittlefair commented:
Hi @adrn,

Excellent work from you, @eteq and @mhvk on implementing velocities. Over the next few days I'll review the PRs in order and give some comments. Here are my broader comments:

1. Like @mhvk I'm a little concerned by the asssumption that the first element of the differential list is the time derivative. Perhaps a dictionary, with the key being the SI unit which the derivative is with respect to would work?

1. Have you considered using ```origin``` instead of ```base``` for ```to_cartesian```?

1. I'd be wary about implementing the operations on representations with differentials attached. It's not immediately obvious to me how I'd expect them to work. One might argue that subtracting differentials with velocities attached ought to implement aberration, for example. On the other hand, raising errors is intrusive. Perhaps a warning is more appropriate?

I'll try and take a finer look at the code over the weekend.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+                                     ' must be a class, not a string. For '
+                                     'strings, use frame objects')
+                if (self.differentials is None or
+                        len(other_class) != 1+len(self.differentials)):
```
For a dict, I think there would be the choice to give just two classes if there is only one differential, or a class for the representation and a dict for the differentials if there is more than one.

adrn replied:
Done, ish

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+        olddiffs = self._differentials
+        try:
+            self._differentials = None
+            newrepr = deepcopy(self)
```
Why a `deepcopy` - since representations are nominally immutable, I think a shallow copy is OK.

But, really, this to me is premature optimization. I think it is better to just use the initializer, with `copy=False` (which means it is not all that slow), and then speed up everything by defining a no-check initializer option (e.g., `copy=None`?); see #5598 for initial work in that direction.

adrn replied:
The way you describe is how I thought we should implement it, but @eteq implemented it this way instead. I'd be ok changing it but I'll let @eteq chime in before making the change.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+
+        # save the differentials, then strip them so that the copy doesn't do
+        # an unneeded copy, then re-attach them to the right objects
+        olddiffs = self._differentials
```
Same comment as above: just use the initializer with `copy=False`.

mhvk replied:
But also: just return `self` if the representation has no differentials; then,
`r1.without_differentials().to_cartesian()` is very fast for that case.

adrn replied:
I implemented your second suggestion, but am waiting for @eteq to chime in on the other comment before swapping for the initializer.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+
+    def to_cartesian(self, diffstocart=False):
+        """
+        Convert the representation to its carteisan form.
```
spelling of cartesian.

adrn replied:
Done

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+        """
+        Convert the representation to its carteisan form.
+
+        Note that subclasses should *not* override this - rather they should
```
This should be a comment in the code (and in BaseRepresentation, where you have it already), not in the docstring.

adrn replied:
Done

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+
+        Parameters
+        ----------
+        diffstocart : bool
```
I don't think this should be an option at all. It should *always* be done. And if it is not possible, the user has to drop the differential beforehand.

This solves the problem of having to define a new function too!

adrn replied:
Done

adrn replied:
Ah, but one side effect of the `diffstocart` kwarg was that it protected against infinite recursion here: `diff.represent_as(CartesianDifferential, base=self)` would again call `to_cartesian()` of the `base` inside the `represent_as`, but with the default `diffstocart=False`. Instead, I had to add remove the differentials when passing in to the `base`... let me know if you can think of a better way.

mhvk replied:
This would disappear if we make `BaseRepresentation.represent_as(<BaseRepresentation>)` drop differentials. Alternatively, we'd have to add `without_differentials()` where appropriate. Will think a bit more.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+            A new representation of this class's type.
+        """
+        repr = cls._from_cartesian_helper(other)
+        repr._differentials = other._differentials
```
I think these need to be transformed here - if the user wants something more complicated, up to them to arrange that outside.

adrn replied:
Transformed to the corresponding differential class, via `DIFFERENTIAL_CLASSES[other.get_name()]`?

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -496,6 +701,33 @@ def from_representation(cls, representation):
+
+        """
+        rep = super(BaseRepresentation, self)._apply(method, *args, **kwargs)
+        diffs = [diff._apply(method, *args, **kwargs)
```
Just go for a tuple directly by "tuple comprehension": `tuple(diff._apply(...) for ...)`

adrn replied:
Made this the same as above, but can change from the wrapped list comprehension if you think it's significantly cleaner without...

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -538,12 +773,24 @@ def _combine_operation(self, op, other, reverse=False):
+    @BaseRepresentationOrDifferential.shape.setter
+    def shape(self, shape):
+        # See: https://stackoverflow.com/questions/3336767/ for an example
+        BaseRepresentationOrDifferential.shape.fset(self, shape)
```
This is trickier: if the differentials have a different broadcast, the shape setting may fail on them. In that case, you have to restore the `shape` of self. So, use a `try/except` as in the `shape.setter` in Base...

Also, always better to use `super`: `super(..., self).shape.fset(self, shape)` (not 100% sure that's possible; if not, don't worry).

adrn replied:
Not sure about the first part: so far, we require that the differentials have the same shape as the representation, so I don't think we'll have broadcasting issues? Of course, it would be nice to allow them to just be broadcastable...

`super()` doesn't work, which is why I had to resort to the class name...not sure if that is the best way to do it though.

mhvk replied:
The problem occurs if, e.g., the representation has a proper (2,5) shape and the differential is really (5,) but broadcast to (2,5). Then you could not change the shape to (5, 2) or even just (10,):
```
In [14]: a=np.broadcast_to(np.ones(5), (2,5))

In [15]: a.shape = 10,
# AttributeError: incompatible shape for a non-contiguous array
```

adrn replied:
Not clear on what you're suggesting - right now, the differentials added to a representation must have the same shape as the representation, so if we just keep that strictness for now we shouldn't have to deal with this?

mhvk replied:
Same shape is not good enough (see above), because unlike `reshape`, the shape setter errors if the reshape is not possible (it has to, since it is inplace). My suggestion is to do the shape-setting on the differentials in a `try/except` loop, and reverse it if it fails anywhere. See the normal `_apply`

adrn replied:
But, for reasons I don't yet understand, this works:

```python
xyz = np.random.random((3, 2, 5)) * u.km
rep = r.CartesianRepresentation(*xyz)

d_x, d_y, d_z = np.random.random((3, 5)) 
dif = r.CartesianDifferential(d_x=np.broadcast_to(d_x, (2,5))* u.km/u.s,
                              d_y=np.broadcast_to(d_y, (2,5))* u.km/u.s,
                              d_z=np.broadcast_to(d_z, (2,5))* u.km/u.s)

rep = rep.with_differentials(dif)

rep.shape = (10,)
```

mhvk replied:
If `with_differentials` is still the version that makes a copy, then indeed it should work -- no strides of 0 any more. I haven't installed your branch yet, but would expect that
```
rep = r.CartesianRepresentation(*xyz, differentials=dif)
rep.shape = (10,)
```
would fail, with `rep.shape` already changed to `(10,)`.

adrn replied:
Yea, strange, this still seems to work:

```python
xyz = np.random.random((3, 2, 5)) * u.km

d_x, d_y, d_z = np.random.random((3, 5)) 
dif = r.CartesianDifferential(d_x=np.broadcast_to(d_x, (2,5))* u.km/u.s,
                              d_y=np.broadcast_to(d_y, (2,5))* u.km/u.s,
                              d_z=np.broadcast_to(d_z, (2,5))* u.km/u.s, 
                              copy=False)

rep = r.CartesianRepresentation(*xyz, differentials=dif, copy=False)
rep.shape = (10,)
```

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -661,7 +913,8 @@ class CartesianRepresentation(BaseRepresentation):
 
-    def __init__(self, x, y=None, z=None, unit=None, xyz_axis=None, copy=True):
+    def __init__(self, x, y=None, z=None, unit=None, differentials=None,
+                 xyz_axis=None, copy=True):
```
I think I'd put `differentials` behind `xyz_axis`, since that really belongs to the coordinates. Also, just for backward compatibility (I worry much less about someone not using `copy=`...)

adrn replied:
Done

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -758,6 +1012,8 @@ def transform(self, matrix):
                        [ 1.23205081, 1.59807621],
                        [ 3.        , 4.        ]] pc>
         """
+        self._raise_if_has_differentials('transform')
```
Is this necessary for `transform`? I think for cartesian, it is reasonable to transform the differentials as well, no? (sort-of assumed so in comments on transformations)

adrn replied:
Ah, I fixed this in the affine transforms PR but will pull those changes in here.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -897,9 +1159,9 @@ class UnitSphericalRepresentation(BaseRepresentation):
+    def __init__(self, lon, lat, differentials=None, copy=True):
         super(UnitSphericalRepresentation,
-              self).__init__(lon, lat, copy=copy)
+              self).__init__(lon, lat, copy=copy, differentials=differentials)
```
Somehow prefer to have differentials before copy here too (I know, it doesn't matter).

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -958,25 +1220,29 @@ def from_cartesian(cls, cart):
-                         self).represent_as(other_class)
+        if inspect.isclass(other_class):
+            if issubclass(other_class, PhysicsSphericalRepresentation):
+                return other_class(phi=self.lon, theta=90 * u.deg - self.lat, r=1.0,
```
This drops the differential! This may actually not be a bad idea, but it is not what the default implementation does.

adrn replied:
Fixed, ish

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
+        """
+        repr = self._to_cartesian_helper()
+        if diffstocart:
+            repr._differentials = tuple([
```
No need to put the list, comprehension also works for a tuple.

adrn replied:
Tuple comprehension doesn't work directly, it only seems to work by doing, e.g.,
```python
tuple(x for x in range(10))
```
because `x for x in range(10)` produces a generator, which then gets turned into a tuple. So it's actually marginally faster to pass in a list comprehension to `tuple()` because the conversion is faster. But we're talking about a 20% speedup at the microsecond level, so I don't really care either way if you think it looks cleaner without the bracket...

mhvk replied:
Learn something new every day! Including that `list([...])` is faster than `list(...)`. Anyway, just leave as is.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1589,28 +1874,48 @@ def _get_base_vectors(cls, base):
     def from_cartesian(cls, other, base):
-        """Interpret 3D rectangular cartesian coordinates as differentials.
+        """
+        Convert the differential to 3D rectangular cartesian coordinates.
```
Bad copy&paste (also, why keep the first line empty? pep-257 asks for the one-line title to be on the same line as the opening triple quotes).

adrn replied:
Yea, I have to remember. I personally don't like that style so my go-to is the line break, but will try to be conscious of the pep.

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1046,6 +1049,16 @@ class LogDRepresentation(BaseRepresentation):
                                         ('logr', u.Dex)])
 
+    # now try passing in a differential
+    diff = CartesianDifferential(d_x=1 * u.km/u.s,
```
I think this test should fail - you should pass in a LogDDifferential

adrn replied:
I tried implementing the `LogDDifferential`, but haven't gotten it to work with the dex units - what unit do I pass in for `d_logd`? See `test_representation.py` in incoming commits. I get "`astropy.units.core.UnitTypeError: Quantity instances require <class 'astropy.units.core.UnitBase'> units, not <class 'astropy.units.function.logarithmic.DexUnit'> instances.`" when I try to initialize a `LogDDifferential`.

mhvk replied:
If you just use the automatic constructor, `DifferentialMeta` will assume that all arguments are `Quantity`, and hence `u.Dex` instances will fail. But those are actually not expected: the unit you should use is `u.dex`, since a difference in `logd` should be unitless:
```
In [10]: u.Dex(1.*u.kpc) - u.Dex(0.5*u.kpc)
Out[10]: <Dex 0.3010299956639812 dex>
```
Of course, for transformations to work, you also need to define `unit_vectors` and `scale_factors` in the class...

If it helps, I can make a PR against current master that tests this.

adrn replied:
I think I correctly implemented `unit_vectors` and `scale_factors` but was just stuck on the dex issues. Just pushed the commits that contain my implementation: https://github.com/adrn/astropy/blob/469c02ef38902b7880638334f656248270989c26/astropy/coordinates/tests/test_representation.py#L1024

adrn replied:
Oh, I see now a possible issue in the `scale_factors`. Anyways, if you want to try your own version off of master I can just rebase on that

mhvk replied:
The problem is in part that your unit vectors are wrong - they should be always have unit length, so no need to include the `d` factor (they should be identical to `SphericalRepresenation.unit_vectors`, so you could just link to that).

For the scale factor, it should just be `1.` -- note that this makes tricky unit verification, the `d^n<comp>/dt^2` implies a difference of `<comp>` which for `LogDRepresentation` does not just have the unit of `log_d`... I'd say we postpone this to later...

adrn replied:
Yes, agreed to postpone. Though I'm not sure I agree with your statement about the unit vectors and scale factor; log-distance is a strange space. In fact, perhaps not the best example for the tests or docs...but let's revisit as a "doc change"

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+        # Only change the representation
+        new_rep = rep1.represent_as(SphericalRepresentation)
+        assert new_rep.get_name() == 'spherical'
+        assert new_rep.differentials[0].get_name() == 'cartesian'
```
Again, don't think this is what should happen: drop or convert.

adrn replied:
Done

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+        matrix = np.array([[1,2,3], [4,5,6], [7,8,9]])
+
+        with pytest.raises(TypeError):
+            s1.transform(matrix)
```
Hopefully no `TypeError` for this!

adrn replied:
Done

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+        assert cr.differentials != cr2.differentials
+        assert cr2.differentials[0] is diff
+
+        # make sure its a copy and not a view
```
Why? (see comment above) Representations are nominally not mutable, so a copy is not needed.

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+        cr2.x.ravel()[:] = 10*u.kpc
+        assert cr.x != cr2.x
+
+        # make sure it works even if a differential is present already
```
Is this what we want? Shouldn't this just error? Or update the differential dict?

adrn replied:
This now updates the differentials dict

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+        assert cr4.differentials[0] != cr3.differentials[0]
+        assert cr4.differentials[0] == diff2
+
+        # also ensure a *scalar* differential will works
```
This won't work with a dict. Unless we have a default for key=None (which we can *not* use ourselves).

adrn replied:
I implemented the `_validate_differentials` method so that if a scalar is passed in, it calls `._get_deriv_key()` to get the appropriate dict key and then loads it into the dictionary. Might be easier than always having to pass in a dictionary?

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1086,208 @@ def test_combine_xyz():
+                                 differentials=[sd])
+
+    notdiff = sr.to_cartesian(diffstocart=False)
+    diff = sr.to_cartesian(diffstocart=True)
```
Again, don't like this - should be done outside of representations if wanted.

adrn replied:
Done

mhvk commented on astropy/coordinates/tests/test_representation_arithmetic.py:
```diff
@@ -1189,3 +1189,36 @@ def test_combinations(self, sd_cls):
         assert np.all(representation_equal(so5, so5c))
         assert_representation_allclose(self.s + (uo+ro), self.s+so1)
+
+@pytest.mark.parametrize('rep', [
```
This iteration would need to change if we don't allow mix-'n'-match

adrn replied:
Done

mhvk commented on astropy/coordinates/transformations.py:
```diff
@@ -519,7 +519,7 @@ def to_networkx_graph(self):
         return nxgraph
 
-    def transform(self, transcls, fromsys, tosys, priority=1):
+    def transform(self, transcls, fromsys, tosys, priority=1, **kwargs):
```
I guess this really belongs in one of the other PRs -- no problem.

adrn replied:
Sorry... I think that got pulled in from a cherry-pick.

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -314,6 +314,138 @@ provides the ``base``::
+   first derivatives, but they do not check the units of the inputs to enforce
+   this. Passing in 2nd derivatives, e.g., acceleration values (with
+   acceleration units), will succeed, but any transformations that occur through
+   re-representation of the differential will not necessarily be correct.
```
Is this true? I think they will be correct (at the representation level at least). They certainly are correct for offsets; I don't think adding another derivative makes a difference. (I did think about this when writing the classes and convinced myself it was OK - that the apparent accelerations appear automatically - but I may have fooled myself).

adrn replied:
Unless I'm spacing here (possible, given the time), don't you need the velocities and the positions to convert the representation of an acceleration? For example, for cylindrical radius:
![image](https://user-images.githubusercontent.com/583379/27249985-ff2f7084-52f0-11e7-958e-ee64dd763578.png)

eteq replied:
I was thinking along the same lines as @adrn here.  My example is spherical coordinates, where you can e.g. look at http://mathworld.wolfram.com/SphericalCoordinates.html and compare (72) to (75)-(80) - that says to me that the spherical<->cartesian acceleration transform has different unit vectors from the velocity transform.

But I admit I did not look *that* closely at the differential machinery to know if it is accounting for this somehow...?

mhvk replied:
Duh, I clearly was too tired yesterday - yes, of course. And, no, the machinery does not (yet) allow that -- I think it will need an `order` keyword or something like that for `scale_factors` - makes me thing that we'd need something information on derivatives - and should get at least that part of the `Variable` work in: https://github.com/astropy/astropy/pull/3715/files#diff-2a3560744629608434d0eff93c50cf2aR13

This also starts me liking more and more @StuartLittlefair's idea of using the unit of the derivative as the key to a `dict` of differentials - that would seem to extend nicely to possible other uses.

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -314,6 +314,138 @@ provides the ``base``::
+.. warning::
+
+    The API for this functionality may change in future versions and should be
+    viewed as experimental!
```
The python term for this seems to be "provisional".

adrn replied:
Done

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -314,6 +314,138 @@ provides the ``base``::
+   differentials=(<CartesianDifferential (d_x, d_y, d_z) in km / s
+      ( 1.,  2.,  3.)>,)>
+
+Implicit in the above is that the representation of a differential does not have
```
Let's not do that! And even if we do that, don't do it in the example...

adrn replied:
Fixed

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -314,6 +314,138 @@ provides the ``base``::
+   differentials=(<CartesianDifferential (d_x, d_y, d_z) in km / s
+      [( 0.,  4.,   8.), ( 1.,  5.,   9.), ( 2.,  6.,  10.), ( 3.,  7.,  11.)]>,)>
+
+However, by passing just the desired representation class, only the
```
Not OK!

adrn replied:
Fixed

mhvk commented:
For a dict of differentials, the question would be whether to allow a single entry. I think so far we've avoided allowing too many short-cuts, so possibly we should do that here too. Alternatively, the key for a single entry could be `None` or an empty string -- with the idea to use the derivative unit as a key, I like the idea of using strings generally as keys (though not insisting on it).

mhvk commented:
p.s. Off to sleep now -- would like to look at the whole lot again -- hope that is OK with those that decide on PR cutoffs.

adrn commented on astropy/coordinates/representation.py:
```diff
@@ -348,11 +361,14 @@ def __repr__(self):
         arrstr = _array2string(self._values, prefix=prefixstr)
 
+        diffstr = ''
+        if hasattr(self, 'differentials') and self.differentials:
```
Done

adrn replied:
Done

adrn commented on astropy/coordinates/representation.py:
```diff
@@ -422,19 +438,74 @@ class BaseRepresentation(BaseRepresentationOrDifferential):
+    differential class, one should also define ``unit_vectors`` and
+    ``scale_factors`` methods (see those methods for details). Finally, classes
+    can also define a ``recommended_units`` dictionary, which maps component
+    names to the units they are best presented to users in (this is used only in representations of coordinates, and may be overridden by frame classes).
```
Done

adrn commented on astropy/coordinates/representation.py:
```diff
@@ -422,19 +438,74 @@ class BaseRepresentation(BaseRepresentationOrDifferential):
 
+    def __init__(self, *args, **kwargs):
+        # now handle any differentials passed in
+        differentials = kwargs.pop('differentials', tuple())
```
Done

adrn commented on astropy/coordinates/representation.py:
```diff
@@ -422,19 +438,74 @@ class BaseRepresentation(BaseRepresentationOrDifferential):
+        ``self._differentials``, but rather leaves that for the caller.
+        """
+        # Now handle the actual validation of any specified differential classes
+        if isinstance(differentials, BaseDifferential):
```
Done

adrn commented on astropy/coordinates/representation.py:
```diff
@@ -483,7 +554,141 @@ def represent_as(self, other_class):
-            return other_class.from_cartesian(self.to_cartesian())
+            if isiterable(other_class):
+                if isinstance(other_class, six.string_types):
+                    raise ValueError('Input to a representation\'s represent_as'
```
Done

adrn commented:
I've implemented changes based on many of @mhvk's smaller comments.

To summarize some of the large changes requested (still need input on some):

- Require that differentials passed in to a representation are "compatible" in
  some sense that I'm still not quite clear on. For some, it is obvious
  (Cartesian, Cylindrical), but for the spherical and unitspherical classes I'm
  less clear on what relations you want to allow and how to do the checks.
  Implementing a `_compatible_differentials` attribute on each representation
  seems a little hacky, but would be a fast way to add what we need, explicitly.

- Switch the `.differentials` attribute from using a list to using a dictionary
  with either (A) integer keys that represent the derivative order (e.g., 1 for
  d/dwhatever, 2 for d/dwhatever^2, etc.), or (B) string keys that are taken 
  from stripping the representation's unit from the differential components. 
  I think (A) would be faster and easier to implement...

- Enable addition and subtraction for representations that contain differentials
  as long as each key in on differential dict has a corresponding member in 
  the other differential dict.

- Change `represent_as` so that passing just a representation class will drop
  the differentials. To transform both, one must specify
  `represent_as(Representation, Differential)` or 
  `represent_as(Representation, dict_of_differentials)`

also cc @eteq @StuartLittleFair

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -348,11 +361,14 @@ def __repr__(self):
         arrstr = _array2string(self._values, prefix=prefixstr)
 
+        diffstr = ''
+        if getattr(self, 'differentials', None):
```
Is there a reason why __str__ ignores differentials when __repr__ does not?

adrn replied:
It wasn't immediately clear to me how we would want `__str__` to look when there are differentials. Also, to me, the repr seems like a more appropriate place to expose more of the internals of the object, so I would actually vote to have them appear in the `__repr__` and not in str, but am open to suggestions.

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -422,19 +438,77 @@ class BaseRepresentation(BaseRepresentationOrDifferential):
+
+        super(BaseRepresentation, self).__init__(*args, **kwargs)
+
+        # store as a private name so users know this is not settable
```
If we really want to enforce non-setting of differentials, it may be best to put the validation logic in the property setter, and raise an exception if ```self._differentials``` exists.

adrn replied:
In the property setter for `.differentials`? Right now you already get an `AttributeError: can't set attribute` if you try to set `.differentials`

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -1193,16 +1489,17 @@ def scale_factors(self, omit_coslat=False):
-        else:
-            return super(SphericalRepresentation,
-                         self).represent_as(other_class)
+        if inspect.isclass(other_class):
```
As @mhvk noted above for UnitSphericalRepresentation, these shortcuts drop the differentials

adrn replied:
Thanks. I put in a solution that just skips the shortcut if a differential is attached, so it will work just not be as efficient when there are differentials. I also added a method that would help us avoid that / keep the optimization, but I'm running out of time, so in the interest of having it work I left TODO's.

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -1340,15 +1639,17 @@ def scale_factors(self):
-            return other_class(lon=self.phi, lat=90 * u.deg - self.theta)
-        else:
-            return super(PhysicsSphericalRepresentation, self).represent_as(other_class)
+        if inspect.isclass(other_class):
```
as before for dropping differentials

StuartLittlefair commented on astropy/coordinates/representation.py:
```diff
@@ -1596,21 +1898,39 @@ def to_cartesian(self, base):
         """
+        return cls._from_cartesian_helper(other, base)
+
+    def _to_cartesian_helper(self, base):
```
I think this function and ```_from_cartesian_helper``` need fuller docstrings to explain to those less familiar with the concepts behind representation of differentials, perhaps with a spherical coordinate or cylindrical coordinate example?

adrn commented:
@mhvk On the idea of `_compatible_differentials`: unfortunately this doesn't work without some major reorganization of the module because the differential classes are defined below the representations. I didn't think of any other easy alternatives, so I could split out the differential classes into a separate module and try to resolve any circular imports...thoughts?

mhvk commented:
@adrn - the `dict` is looking good. I guess this doesn't decide on the key yet. Options: integer, string. Initially I had for the latter `1/s`, etc., but maybe since it is for differentials it is more logical to do `s1`, `s2`, i.e., the unit of the denominator. Shorter yet equally clear.

I also like the change to have `represent_as` to drop differentials unless a relevant class is given. 

But I think the latter, as well as the belated realization that one cannot "just convert to cartesian" for higher order, makes me feel that we should, unlike what I originally thought, drop differentials a bit more easily. In particular:
* I think `to_cartesian()` should also drop the differentials (I don't like the many `without_differentials()` I'm seeing... and really it should do the same as `represent_as(CartesianRepresentation)`). To keep the differntials, one would then have to go through `represent_as`. But at least we remain backward compatible. Eventually, perhaps we could add a `keep_differentials` argument (`False` by default) which would then do the correct transformation of the differentials. But we do not need that now.
* Continuing from this thought, I also think methods like `norm` and `dot`, which do not return a representation, should continue to work as before (I was sad to see you had to circumvent `norm` somewhere).
* I'm still not sure still about operations that return a representation, though. I *think* I'd like those to work on differentials too (and on first sight, it seems well-defined for cartesian, at least for scaling and subtraction/addition). So, perhaps continue to fail on those for now just to not set a precedent that we cannot undo.

adrn commented:
@mhvk OK, I generally agree with your comments so will work on that now.

adrn commented:
@mhvk And just to be sure: you were suggesting that, if we go back to dropping the differentials with `to_cartesian()`, we could do away with the `_helper()` methods and thus remain backwards-compatible with subclasses implementing `to/from_cartesian`?

mhvk commented:
@adrn - yes, that's one advantage - we don't need the new helper functions.

adrn commented:
OK, I think I've gone through all of your comments @mhvk and @StuartLittleFair. I'll ping @eteq on the few questions we have for him. Thanks much for taking a look at this!

At this point, I'd like to ask for some help: if you have major comments/changes to request, leave them here. If you have a significant number of minor comments, could you please submit via PR to this branch on my fork? I'm running out of time myself, and we still have two more PR's to go through...

Otherwise, here are some misc. notes:

* The dictionary keys are now simply the SI derivative unit, e.g., 's' for a first time derivative.
* (most relevant to @eteq who added it) I removed the `diffstocart` option in `BaseRepresentation.to_cartesian()` because it now drops the differentials. To convert to cartesian, you have to go through `represent_as(CartesianRepresentation, CartesianDifferential)`

Questions and remaining TODO's:
* [ ] Decide on what to do with `with_`/`without_differentials` - possibly axe the deepcopy and just go through the initializer (@eteq)
* [ ] The logic in `BaseCoordinateFrame._re_represent_differentials()` feels a bit clunky to me - can someone take a look and see if you would do anything to improve? 
* [ ] The tests are failing (so I canceled them) because of the implementation of `LogDDifferential` in `test_representation.py`. I'm hoping @mhvk can just suggest to me the changes - I had to move on because I'm stuck.
* [ ] Do we want to include the differentials in `__str__` or just `__repr__`?
* [ ] Re-enable (at least) addition and subtraction of representations that have differentials. 
    * [ ] One remaining unknown is what to do in the case that `rep1+rep2` only `rep1` has differentials? Raise an error?
* [ ] Update documentation

eteq commented:
> (most relevant to @eteq who added it) I removed the diffstocart option in BaseRepresentation.to_cartesian() because it now drops the differentials. To convert to cartesian, you have to go through represent_as(CartesianRepresentation, CartesianDifferential)

I didn't fully follow the conversation above about that (can review more if its critical), but this is very problematic to me.  I have a *lot* of lines that I've been writing that now turn into something like:
```
frame.data.represent_as([CartesianRepresentation, CartesianDifferential]).differentials[0]
```
for what seems to me to be the single *most important* thing that representations need to be able to do - that is, the representation framework is built around the idea that "everything goes through cartesian".  And that's true for a lot of the transforms, where you need to switch to cartesian to do operations.

Maybe the workaround is to add a `to_cartesian_differentials()` method?   Does that seem good to @mhvk too?  If so I can do that to releive @adrn's burden some.


> Decide on what to do with with_/without_differentials - possibly axe the deepcopy and just go through the initializer 
 
(Assuming this is about https://github.com/astropy/astropy/pull/6169#discussion_r122533620) I'm ok with going through the initializer.  The reason for the implementation was just what @mhvk pointed out of trying to optimize out all the initializer checks... The `deepcopy` vs. `copy` was based on the idea that I *fear* people are violating the immutability.  But I see the point that this is being a bit too fearful so if you both agree initializer is better, then 👍 

> Do we want to include the differentials in __str__ or just __repr__

I would vote for something like this:
* `__str__` yields: `"( 0.,  0.,  10.) (deg, deg, kpc)"`
* `__repr__` yields:
```
'<SphericalRepresentation (lon, lat, distance) in (deg, deg, AU) with differentials=[SphericalDifferential]
    ( 12.3,  45.6,  7.8)'
```
The current `__repr__` showing the values of the differentials makes for something *very* hard to read, so this version just shows that they exist.  User then has to do `r.differentials` to get the actual differential `__repr__`'s and such.  But that makes it much less cluttered.

> One remaining unknown is what to do in the case that rep1+rep2 only rep1 has differentials? Raise an error?

I vote either 1) drop the differential or 2) treat it as (0,0,0) .  I slightly favor 1 as it seems less likely to silently lead to errors, but could go either way.  Am against error because that would make it harder to do operations on representations when you don't care about the differentials (but they were attached in some other context for some other reason)

mhvk commented:
@eteq - am too sleepy really but still up: if you know you want the differential, you can of course transform *it* to cartesian without problem: `r.differentials['s'].to_cartesian(base=r)`. My main problem is that we now realized `to_cartesian` would do the wrong thing for higher order, and I don't think we should make that easy... 

p.s. An alternative would be a default-False `keep_differentials` keyword argument - this would then fail on anyone who defined subclasses (think sunpy) once one included differentials.

adrn commented:
@mhvk @eteq just mentioned the same idea to me in chat, but I slightly favor having a separate method rather than baking it in to `to_cartesian()` - then we don't have to worry about inconsistencies with subclasses that forget to or choose not to support `keep_differentials`.

adrn commented:
For `__str__` I agree, no differential info is shown. For `__repr__`, I went with showing a visual list of the differential units (not the classes), e.g.:
```
<CartesianRepresentation (x, y, z) in pc
    [( 0.14269309,  0.81152416,  0.83854298),
     ( 0.78899734,  0.73822164,  0.90117812),
     ( 0.23232916,  0.62093174,  0.49368429),
     ( 0.52731394,  0.16454857,  0.85167247),
     ( 0.46427483,  0.17689298,  0.84388302)]
 differentials=[1/s, 1/s2]>
```

adrn commented:
@mhvk @eteq OK, I think now I've addressed all points I could find in this PR, so is now ready for another review. I'd especially like to settle on any API decisions with this PR ASAP so I can start rebasing the other branches.

Currently, even simple combine operations (add, subtract, ...) with representations that contain differentials will error. I'm fine leaving this as is until we've thought through how to do the operations unambiguously, but will wait to hear what you think.

Still also waiting to see if @eteq wants (or wants to implement) the extra `to_cartesian_...` method.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -478,12 +612,74 @@ def represent_as(self, other_class):
+        Returns
+        -------
+        newrepr
+            A copy of this representation, but without the ``differentials``.
```
Maybe explicitly say "shallow copy"? In any case, definitely state that one gets the original back if it has no differentials attached (which I think is correct behaviour).

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -496,6 +692,34 @@ def from_representation(cls, representation):
+
+        rep._differentials = dict(
+            [(k, diff._apply(method, *args, **kwargs))
+             for k,diff in self._differentials.items()])
```
nitpick: space after `,`

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -538,18 +765,42 @@ def _combine_operation(self, op, other, reverse=False):
+        try:
+            for k in self.differentials:
+                self.differentials[k].shape = shape
+        except:
```
`except Exception`

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -718,9 +977,32 @@ def get_xyz(self, xyz_axis=0):
 
     @classmethod
     def from_cartesian(cls, other):
+        """
```
Since we're using `InheritDocStrings`, maybe just put the docstring on the very base?

Note that we're not converting here, so if we add a specific docstring to `Cartesian` it should say so.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -718,9 +977,32 @@ def get_xyz(self, xyz_axis=0):
         return other
 
     def to_cartesian(self):
+        """
```
Same.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -784,10 +1067,26 @@ def transform(self, matrix):
+        # Handle differentials attached to this representation
+        diffs = None
+        if self.differentials:
+            # We might be being too generous here - we could enforce that this
```
Now that for this case we *know* these are `CartesianDifferentials`, we can simplify this.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1097,9 +1422,13 @@ def from_cartesian(cls, cart):
         Converts 3D rectangular cartesian coordinates to radial coordinate.
         """
-        return cls(distance=cart.norm(), copy=False)
+        # We can't use .norm() here because it's not differentials-safe
```
This is no longer true, so we can revert this.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1247,6 +1588,7 @@ def norm(self):
         norm : `astropy.units.Quantity`
             Vector norm, with the same shape as the representation.
         """
+        self._raise_if_has_differentials('norm')
```
Norm is OK, so remove.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1393,6 +1741,7 @@ def norm(self):
         norm : `astropy.units.Quantity`
             Vector norm, with the same shape as the representation.
         """
+        self._raise_if_has_differentials('norm')
```
Norm is OK, so remove.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -991,10 +1309,13 @@ def norm(self):
         norm : `~astropy.units.Quantity`
             Dimensionless ones, with the same shape as the representation.
         """
+        self._raise_if_has_differentials('norm')
```
can be removed.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1112,6 +1441,7 @@ def norm(self):
         norm : `~astropy.units.Quantity`
             Dimensionless ones, with the same shape as the representation.
         """
+        self._raise_if_has_differentials('norm')
```
can be removed.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1564,6 +1913,23 @@ def _check_base(cls, base):
+
+        # TODO: better validation by checking, e.g., _compatible_differentials
+        # for this class
+        self._check_base(base)
```
I think this may indeed be too strict, but we can revisit if it turns out to be a problem.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -1564,6 +1913,23 @@ def _check_base(cls, base):
+            d_comp = getattr(self, 'd_{0}'.format(name), None)
+            if d_comp:
+                d_unit = comp.si.unit / d_comp.si.unit
+                return str(d_unit)
```
I think this is too inprecise:
```
str(u.deg.si/(u.mas/u.yr).si)
# '1.13607e+14 s'
```
I think we just want the SI base and power. The easiest is a bit of a trick, but it works:
```
str(u.Quantity(1., comp.unit / d_comp.unit).si.unit)
str(u.Quantity(1., u.deg/(u.mas/u.yr**2)).si.unit)
# 's2'
```

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1080,239 @@ def test_combine_xyz():
+                                     d_z=3 * u.km/u.s)
+
+        # can pass in a single differential, which gets turned into a
+        #   length-1 tuple
```
tuple -> dict

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1080,239 @@ def test_combine_xyz():
+
+        r2 = CartesianRepresentation.from_representation(r1)
+        assert r2.get_name() == 'cartesian'
+        # assert r2.differentials['s'].get_name() == 'cartesian'
```
delete

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1073,3 +1080,239 @@ def test_combine_xyz():
+def test_to_cartesian():
+    """
+    Test that to_cartesian does the expected thing (converts both the
+    representation and differentials)
```
Update docstring.

mhvk commented:
OK, pushed my changes, including a second commit with also `__repr__` (and a bug fix of the first commit, so be careful if one wants to revert that).

I would suggest that if we want to have a method that transforms differentials to cartesian as well, we discuss that in another PR, so that we can merge this one.

adrn commented:
Thanks @mhvk for the careful reading and for submitting changes! From your perspective, is this ready to go if tests all pass? I'd love to have the satisfaction of hitting merge on this one 😄 , but will wait until @eteq chimes in to see if he has any remaining comments.

mhvk commented:
Yes, I'm :+1: on merging. This has become a really nice integration of the differentials!!

Next up is #6218, correct? That one I think was very close as it was; hopefully the rebase will not be too painful.

adrn commented:
Yes, next is #6218 - talk to you over there!

adrn commented:
(Remote data build failures)

eteq commented on CHANGES.rst:
```diff
@@ -20,12 +20,16 @@ New Features
     coordinate frames that allow copying an existing frame object with various
     reference or copy behaviors and possibly overriding frame attributes. [#6182]
 
+  - The ``Representation`` class instances can now contain ``Differential``
```
I think these shouldn't be in backticks because they're not actual class names... Could either do `BaseRepresentation` in backticks or just leave off the backticks and lower-case them to not be referncing sepcific class names

adrn replied:
Oops - I had meant to do ``*Representation`` but I'll just switch to plain text, lowercase.

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -661,7 +938,8 @@ class CartesianRepresentation(BaseRepresentation):
                                 ('z', u.Quantity)])
 
-    def __init__(self, x, y=None, z=None, unit=None, xyz_axis=None, copy=True):
+    def __init__(self, x, y=None, z=None, unit=None, xyz_axis=None,
```
The docstring doesn't have "differnentials" mentioned...

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -897,9 +1194,15 @@ class UnitSphericalRepresentation(BaseRepresentation):
         return SphericalRepresentation
 
-    def __init__(self, lon, lat, copy=True):
+    def __init__(self, lon, lat, differentials=None, copy=True):
```
again, no `differentials` in the docstring

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -1146,13 +1463,20 @@ class SphericalRepresentation(BaseRepresentation):
     _unit_representation = UnitSphericalRepresentation
 
-    def __init__(self, lon, lat, distance, copy=True):
+    def __init__(self, lon, lat, distance, differentials=None, copy=True):
```
docstring needs `differentials`...

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -1279,9 +1607,10 @@ class PhysicsSphericalRepresentation(BaseRepresentation):
     recommended_units = {'phi': u.deg, 'theta': u.deg}
 
-    def __init__(self, phi, theta, r, copy=True):
+    def __init__(self, phi, theta, r, differentials=None, copy=True):
```
docstring needs `differentials`...

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -1068,8 +1383,9 @@ class RadialRepresentation(BaseRepresentation):
 
-    def __init__(self, distance, copy=True):
-        super(RadialRepresentation, self).__init__(distance, copy=copy)
+    def __init__(self, distance, differentials=None, copy=True):
```
docstring needs `differentials`...

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -1422,9 +1756,10 @@ class CylindricalRepresentation(BaseRepresentation):
     recommended_units = {'phi': u.deg}
 
-    def __init__(self, rho, phi, z, copy=True):
+    def __init__(self, rho, phi, z, differentials=None, copy=True):
```
docstring needs `differentials`...

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -435,6 +467,78 @@ class BaseRepresentation(BaseRepresentationOrDifferential):
 
     recommended_units = {}  # subclasses can override
 
+    def __init__(self, *args, **kwargs):
```
More debatable here, but it would probably be useful for this to have a docstring of what's valid for ``differentials`` here

eteq commented on astropy/coordinates/representation.py:
```diff
@@ -348,11 +376,15 @@ def __repr__(self):
+        diffstr = ''
+        if getattr(self, 'differentials', None):
+            diffstr = '\n (has differentials: {0})'.format(
+                ', '.join([repr(key) for key in self.differentials.keys()]))
```
I find the description here confusing because it implies the differentials *have that unit*.  How about "has differentials: d/d[s]" or something like that to make it clear its the *derivative* unit?    I think this might do it:
```
', '.join(['d/d[{}]'.format(key) for key in ks])
```

adrn replied:
Yea, this is why I ended up agreeing with you about just showing 'has differentials'. What is shown now are the *keys* to the differentials dictionary. I'm not sure how to get rid of the ambiguity you mention - I worry that if we do `d/ds` people might get confused about what the actual key is. I think we either do something like `has differentials with keys=['s', 's2']` or just go back to `has differentials`

adrn replied:
cc @mhvk FYI: this comment is related to why I had removed exposing the units from the `__repr__`

eteq replied:
It's clear there's not consensus here, so I'm ok with merging as it stands on this issue and will make a separate small PR to make the change I suggested.

eteq commented:
One more suggestion after some out-of-band questioning with @adrn : there should be a test to ensure that the following is *not* accepted as a differential ``{u.s: some_diff, u.yr: some_other_diff}``.  That shouldn't work b/c it's the same physical type so its ambiguous what's going on.  @adrn thinks that's already the case, but is going to add a test.

adrn commented:
The new test I added passes locally, but since I added quite a few lines of documentation I am trying out the new `[docs only]` commit message flag to run the documentation builds.

adrn commented:
Hm, it seems that the `[docs only]` flag didn't work, so I manually killed all tests except for the docs build on travis (and circleci since it's usually fast).
cc @bsipocz

bsipocz commented:
Hmm, what do you mean not working? I think it was: https://travis-ci.org/astropy/astropy/jobs/244281543#L381

(Sadly we can't avoid to start up the jobs, but can just cancel them before any conda install and the testing happens.)

adrn commented:
Oh, ignore my comment! I saw the jobs spinning up and jumped to conclusions. I take it back!

eteq commented:
I'm satisfied, so when the doc tests pass feel free to pull the trigger @adrn .  You've definitely earned that! 💯

adrn commented:
Doh - the docs build failure is because of 8 warnings about mistaken single backticks I had in the docstrings I added. I fixed and ran `build_docs` locally to verify those warnings went away.

adrn commented:
Thanks all for the review -- moving on to #6218

eteq commented:
🚀 🤘 👏 

🔁
